sonar-plugin-api/src/main/java/org/sonar/api/server/rule/StringPatternValidator.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Sonar Plugin API
+ * Copyright (C) 2009-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.api.server.rule;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Set;
+import org.apache.commons.lang.StringUtils;
+
+import static java.lang.String.format;
+import static java.lang.String.join;
+import static java.util.Locale.ENGLISH;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * Validates strings based on the defined regex. The error message is based on the error message passed in parameter.
+ */
+public class StringPatternValidator {
+
+  private final String validCharacterRegex;
+  private final String errorMessageSuffix;
+
+  public StringPatternValidator(String validCharacterRegex, String errorMessageSuffix) {
+    this.validCharacterRegex = validCharacterRegex;
+    this.errorMessageSuffix = errorMessageSuffix;
+  }
+
+  public boolean isValid(String entry) {
+    return StringUtils.isNotBlank(entry) && entry.matches(validCharacterRegex);
+  }
+
+  public String validate(String entry) {
+    if (!isValid(entry)) {
+      throw new IllegalArgumentException(format("Entry '%s' is invalid. %s", entry, errorMessageSuffix));
+    }
+    return entry;
+  }
+
+  public Set<String> validate(Collection<String> entries) {
+    Set<String> sanitizedEntries = entries.stream()
+      .filter(Objects::nonNull)
+      .filter(entry -> !entry.isEmpty())
+      .map(entry -> entry.toLowerCase(ENGLISH))
+      .collect(toSet());
+    Set<String> invalidEntries = sanitizedEntries.stream()
+      .filter(entry -> !isValid(entry))
+      .collect(toSet());
```

Comment:
Remarks:
1. I don't expect a method called validate to produce as a side effet a sanitized set of string
2. this method was never used inside the plugin-api; but it might be used externally so we probably need to keep the behaviour
3. I would simplify this method to:
```
  public static void validate(Collection<String> tags) {
    String invalidTags =   tags.stream()
      .filter(RuleTagFormat::isValid)
      .collect(Collectors.joining(","));
    if (!invalidTags.isEmpty()) {
      throw new IllegalArgumentException(format("Tags '%s' are invalid. %s", join(", ", invalidTags), ERROR_MESSAGE_SUFFIX));
    }
  }
```

and update RuleTagFormat::validate to do the sanitization their if you want to preserve backward-compatibility

Reply:
Indeed, I think it is better to have in our new method/Validator a cleaner way to validate without actually changing the keys, and I think it is not actually needed for the generic concept (we prefer to fail if the generic concept is malformed).
I have made the changes so for the tags it keeps sanitizing, and for the generic concept it is just validating